### PR TITLE
Update to CircleCI 2.0

### LINKED ISSUE #67: CircleCI 2.0 Example
CircleCI 2.0 is in beta testing currently. I have access to it and have found it incredibly useful for a few repos so far where setting up the required external dependencies would have proven _very_ difficult on either Travis-CI or CircleCI's default images.

I would just submit a PR with an example configuration, but I'm not sure how to do so without confusing people new to CI (the supposed target of this repo) by having two different configurations for CircleCI, one of which likely doesn't work for them.

Does anyone have a good idea on how to handle this? Will a clear README be sufficient?

### PR DESCRIPTION
**Pull request by Arcanemagus**
### Description of the Change

* Update the recommended configuration to a basic CircleCI 2.0 configuration.
* Changes the build script to assume unknown CircleCI images are Ubuntu based (as all CircleCI provided Linux images are).

### Alternate Designs

There are two different aspects that could be changed. Note that these are _not_ mutually exclusive, and in fact work better together 😉.

#### Custom Docker Image

Currently this configuration mimics the way that Travis CI and AppVeyor work in that it installs Atom on the fly, an alternative possible with the move to CircleCI 2.0 is the use of a custom Docker image that has Atom pre-installed, allowing that entire process to be done before the builds, leading to a potential significant speed up. On hosts that don't have the image cached downloading the image takes ~25 seconds (installation usually takes ~90), if you get lucky and the host already has the image cached the _entire build_ could be around 10 seconds.

The downside of this is that somebody has to maintain a Docker image. I'm maintaining one for my own builds at [arcanemagus/atom-docker-ci](https://hub.docker.com/r/arcanemagus/atom-docker-ci/), which provides both `stable` and `beta` tags, as well as specific tags for every version (back to v1.14.0 currently). If you are okay with utilizing this, the extra package installation step can be removed entirely, and the `Xvfb` step can be greatly simplified.

There is a downside to this with regards to how the build script is currently written: Right now it assumes it must do _all_ parts of the CI process, including installing Atom. If this is the way you want to move forward the `build-package.sh` script should be split up into a part that installs Atom, and a part that actually runs the specs and lints the code (or split into three parts?). For backwards compatibility the `build-package.sh` script should of course be written to simply run all parts in order, but newer builds can only use the parts that they need.

#### Workflows

CircleCI 2.0 supports what they call "Workflows" now, essentially splitting your build into sections that can depend on each other and can run in parallel. This is somewhat similar to the matrix builds from Travis CI, but far more flexible. An example of how this can be put to use can be found at [here](https://circleci.com/workflow-run/1a404387-32c4-4b36-9eb9-9d6973d9c51c), but it can be pretty easily summed up with a picture:
image

### Why Should This Be In Core?

CircleCI 1.0 configurations like the currently recommended template will cease to function on 2018-08-31, as such the configuration **must** be upgraded before then.

### Benefits

The recommended CircleCI configuration will continue to work past 2018-08-31.

### Possible Drawbacks

* The configuration file is more complex than the 1.0 configuration file, leading to the potential for new users to be confused by what each part is doing.
* The build script will now assume that _any_ build on CircleCI that doesn't specify `osx` in the `CIRCLE_BUILD_IMAGE` environment variable is an Ubuntu based image. If users later on customize the Docker image used to something that breaks this assumption, they will need to modify the script or replace it.

### Verification Process

An example build using this configuration can be found here: [linter-bootlint@305](https://circleci.com/gh/AtomLinter/linter-bootlint/305).

### Applicable Issues

Fixes #67.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    docker:
+    - image: circleci/node:latest
+    steps:
+    - checkout
+    - run:
+        name: Update system package lists
+        command: sudo apt-get update
+    - run:
+        name: Install some pre-requisite packages
+        command: sudo apt-get --assume-yes --quiet install curl libxss1 libx11-xcb1 libasound2 xvfb
```
Comment (Arcanemagus) [inline .circleci/config.yml]: `libxss1 libx11-xcb1 libasound2` can all be removed from here once https://github.com/atom/atom/pull/16812 gets merged and released on `stable`.

Comment (50Wliu) [reply .circleci/config.yml]: Can a comment be added to the file itself saying the same?

Comment (Arcanemagus) [reply .circleci/config.yml]: Done 😉.

**Comment by smashwilson** [comment]
👍 Excellent, thanks 🙇 

> The downside of this is that somebody has to maintain a Docker image. I'm maintaining one for my own builds at arcanemagus/atom-docker-ci, which provides both stable and beta tags, as well as specific tags for every version (back to v1.14.0 currently). If you are okay with utilizing this, the extra package installation step can be removed entirely, and the Xvfb step can be greatly simplified.

Ooh, fancy. Are you using some kind of automation to build this, or are you pushing new tags manually whenever we release?

> If this is the way you want to move forward the `build-package.sh` script should be split up into a part that installs Atom, and a part that actually runs the specs and lints the code (or split into three parts?). For backwards compatibility the `build-package.sh` script should of course be written to simply run all parts in order, but newer builds can only use the parts that they need.

The build template would need to modified to download and `chmod +x` all of the component scripts then, yeah - or do it with a tarball instead... ? Breaking `build-package.sh` into pieces would require us to take a slightly different approach to the distribution process.

In general I'm 👍 on doing something like this, but it'd have to be done with some care to avoid breaking all of the packages that are curling `build-package.sh` from master. (I... have totally done this before 😅 )

**Comment by smashwilson** [comment]
Also: follow-on work here is to upgrade all of our atom/ packages that are using Circle 1.0 to using the new template. I'm on reactive duty this week, so I'll tackle it for any packages that I'm doing other work in 🔧

**Comment by smashwilson** [comment]
Oh yeah, and:

> The build script will now assume that any build on CircleCI that doesn't specify osx in the CIRCLE_BUILD_IMAGE environment variable is an Ubuntu based image.

When I was upgrading atom/github to 2.0, I had to [specify `CIRCLE_BUILD_IMAGE: osx` myself](https://github.com/atom/github/blob/master/.circleci/config.yml#L17) - it didn't look like the 2.0 MacOS image set that variable any more.

**Comment by Arcanemagus** [comment]
> Ooh, fancy. Are you using some kind of automation to build this, or are you pushing new tags manually whenever we release?

I just run https://github.com/Arcanemagus/atom-docker-ci/blob/master/Update-Version.ps1 manually on every release, I haven't yet looked into whether there is a way for external utilities can hook into the releases but it's on my radar as something to do 😉.

> it'd have to be done with some care to avoid breaking all of the packages that are curling `build-package.sh` from master.

Yea, that's the big concern with that project. Since it will provide such a nice simplification of the workflow style builds I definitely plan on trying it out on my fork of `ci` with some of the AtomLinter repos.

> When I was upgrading atom/github to 2.0, I had to [specify `CIRCLE_BUILD_IMAGE: osx` myself](https://github.com/atom/github/blob/master/.circleci/config.yml#L17) - it didn't look like the 2.0 MacOS image set that variable any more.

Yea, none of their 2.0 images do, I'm guessing they just wanted to move away from that for some reason.

**Pull request merged**